### ISSUE TITLE
False positive using PreferMethodReference

### ISSUE THREAD

**Issue by jdussouillez** [sevntu-checkstyle/sevntu.checkstyle#919]
We get some false positive using rule `PreferMethodReference`. This is a specific case when you can't use method reference because of ambiguous context because of overloaded methods for example.

Test.java :
```java
import java.util.function.Consumer;

public class Test {

    public static void main(final String[] args) {
        var user = new User();
        var wrapper = new Wrapper();

        // What Checkstyle is not happy about
        wrapper.consume(t -> user.type(t));

        // What Checkstyle want us to do, but it doesn't compile: "both method consume(Consumer<String>) in Wrapper and method consume(Runnable) in Wrapper match"
        // wrapper.consume(user::type);

        // Our current workaround (do not forget SuppressWithPlainTextCommentFilter module in the config)
        // CHECKSTYLE:OFF: PreferMethodReference
        wrapper.consume(t -> user.type(t));
        // CHECKSTYLE:ON: PreferMethodReference
    }

    private static class Wrapper {

        public void consume(final Consumer<String> consumer) {
            consumer.accept("foo");
        }

        public void consume(final Runnable runnable) {
            runnable.run();
        }
    }

    private static class User {

        private String type;

        public String type() {
            return type;
        }

        public User type(final String type) {
            this.type = type;
            return this;
        }
    }
}
```

Checkstyle.xml :

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN" "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="error"/>
    <property name="fileExtensions" value="java"/>
    <module name="SuppressWithPlainTextCommentFilter"/>
    <module name="TreeWalker">
        <module name="PreferMethodReference"/>
    </module>
</module>
```

```sh
$ java -cp sevntu-checks-1.42.0.jar:checkstyle-10.3.4-all.jar com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml Test.java

Starting audit...
[ERROR] /.../Test.java:10:27: Lambda can be replaced with method reference. [PreferMethodReference]
Audit done.
Checkstyle ends with 1 errors.
```

----

- Checkstyle [10.3.4](https://github.com/checkstyle/checkstyle/releases/tag/checkstyle-10.3.4)
- Sventu checkstyle [1.42.0](https://github.com/sevntu-checkstyle/sevntu.checkstyle/releases/tag/1.42.0)
- Java :
```
openjdk version "17.0.4" 2022-07-19
OpenJDK Runtime Environment (build 17.0.4+8-Ubuntu-120.04)
OpenJDK 64-Bit Server VM (build 17.0.4+8-Ubuntu-120.04, mixed mode, sharing)
```

**Comment by romani** [comment]
>This is a specific case when you can't use method reference because of ambiguous context (because of overloaded methods for example).

Yes, we can not anything in this case. Checkstyle is not type aware tool, and no access to other files (whole validation is done based on code of single file)
https://checkstyle.org/writingchecks.html#Limitations

**Comment by romani** [comment]
I can suggest to use more advanced filters/suppressions to make it less aggressive in code to work around such problems: https://checkstyle.org/config_filters.html attention to xpath, and comments based with "influence", single suppression modules ( to keep suppressions in config , not in code - no pollution in code from tool)
